zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,550 @@
+                __nonempty: (),
+            }
+
+            #[repr(C)]
+            struct #raw_ident #impl_generics #where_clause {
+                pub __tag: #discriminant_ident,
+                pub fields: #fields_ident #ty_generics,
+            }
+
+            // SAFETY:
+            // - `cast` is implemented as required.
+            // - By definition, `*mut Self` and `*mut [u8; size_of::<Self>()]`
+            //   are types of the same size.
+            let discriminant = unsafe {
+                candidate.cast_unsized(|p: *mut Self| {
+                    p as *mut [
+                        core_reexport::primitive::u8;
+                        core_reexport::mem::size_of::<#discriminant_ident>()
+                    ]
+                })
+            };
+            // SAFETY: Since `candidate` has the invariant `Initialized`, we
+            // know that `candidate`'s referent (and thus `discriminant`'s
+            // referent) is as-initialized as `Self`. Since all of the allowed
+            // `repr`s are types for which all bytes are always initialized, we
+            // know that `discriminant`'s referent has all of its bytes
+            // initialized. Since `[u8; N]`'s validity invariant is just that
+            // all of its bytes are initialized, we know that `discriminant`'s
+            // referent is bit-valid.
+            let discriminant = unsafe { discriminant.assume_valid() };
+            let discriminant = discriminant.read_unaligned();
+
+            // SAFETY:
+            // - `cast_unsized`:
+            //   - The raw enum has the same size as the input enum
+            //   - The returned pointer is cast from `p`, and so has the same
+            //     provenance as `p`.
+            //   - The raw enum has the same types at the same locations as the
+            //     original enum, and so preserves the locations of any
+            //     `UnsafeCell`s.
+            // - `assume_initialized`: `cast_unsized` removes the
+            //   initialization invariant from `p`, so we re-assert that
+            //   all of the bytes are initialized.
+            // - `project`:
+            //   - This projection returns a subfield of `this` using
+            //     `addr_of_mut!`.
+            //   - Because the subfield pointer is derived from `this`, it has
+            //     the same provenance.
+            //   - The locations of `UnsafeCell`s in the subfield match the
+            //     locations of `UnsafeCell`s in `this`. This is because the
+            //     subfield pointer just points to a smaller portion of the
+            //     overall struct.
+            let fields = unsafe {
+                candidate
+                    .cast_unsized(|p: *mut Self| p as *mut #raw_ident)
+                    .assume_initialized()
+                    .project(|p: *mut #raw_ident| {
+                        core_reexport::ptr::addr_of_mut!((*p).fields)
+                    })
+            };
```

Comment:
Can you split this into multiple `unsafe` blocks that have separate safety comments?

Reply:
Done